### ISSUE TITLE
Many to Many behaviour

### ISSUE THREAD

**Issue by firstactivemedia** [phalcon/cphalcon#2871]
The way many to many behaves, does not appear to be very intuitive, and is not nice to use. for example

```
$tag1 = \Tag::findFirst('id=1');
$tag2 = \Tag::findFirst('id=2');
$tag3 = \Tag::findFirst('id=3');

// create a post
$post = new \Post();
$post->tags = array($tag1, $tag2);
$post->save();

// edit a post
$post = \Post::findFirst('id=1'); // the same post that was created earlier
$post->tags = array($tag1, $tag3);
$post->save();
```

Will tend to mean the post ends up with all three tags, possibly even with an extra copy of tag1, rather than just 2 as might be expected, and how Doctrine works.

Is this a bug, or intended behaviour?

Some official word on this would be great, seems to attract a few people asking about it http://forum.phalconphp.com/discussion/2190/many-to-many-expected-behaviour

thanks :)
##

**Comment by Pajamaman** [comment]
@phalcon said:

> Not a bug, but probably something that can be improved with an array_unique

Seems like a bug to me.

**Comment by sergeyklay** [comment]
@Pajamaman +1

**Comment by fl0pp** [comment]
Have someone come up with a solution to this issue? It will do with plain PHP-code...

**Comment by prodigga** [comment]
This small issue steered me away from Phalcon a year ago.

**Comment by Jurigag** [comment]
Well i will look into it when i have time, but to be honest i created service myself to create/update many to many relation.

**Comment by fl0pp** [comment]
@Jurigag could you please share that code with me? I consider doing it myself, but if you have already done it, I would be happy to use it!

**Comment by fl0pp** [comment]
I created the following method that I placed in my base model: http://pastebin.com/MikKbSFj

**Comment by makerlabs** [comment]
Wouldn't it be easier to just remove all the relations between the current post and tags and then to add them again this is how I normally deal with this kind of problem when it comes to many-to-many relations. Even when u look from the UI side its always either show all the related tags already added with option to add another one so either u use ajax and add only the one or two new tags or submit all the tags with a normal request again and use the method which I described above.

**Comment by firstactivemedia** [comment]
> Wouldn't it be easier to just remove all the relations between the current post and tags and then to add them again

The original post was more intended to find out of Phalcon is working as intended. Seems the official line is that it is, so no bug...?

Unsure if I have been spoilt by Doctrine doing more for me than phalcon, or fallen into some kind of anti-pattern / over-reliance on the ORM!

**Comment by angelvega93** [comment]
> I created the following method that I placed in my base model: http://pastebin.com/MikKbSFj

It works fine :+1: but it would be even better if you could do like laravel 5

$article->sync('categories', [1, 2 => ['hidden' => 1], 3])

**Comment by makerlabs** [comment]
@firstactivemedia sorry I was just stating a resolution which @phalcon provided I wasn't talking about if its a bug or not :) But when u talk about object oriented approach than yeah I totally agree about the fact that if you assign an array to a object property it should be overriten like it would be in normal object.

**Comment by StudioMaX** [comment]
This is still relevant.

**Comment by StudioMaX** [comment]
Still doesn't work.

**Comment by prodigga** [comment]
Lol. So, is phalcon dead?

**Comment by niden** [comment]
Not at all. We just don't have a lot of resources to fix everything. Some of the issues have been opened a long time ago and we have a bot that closes them automatically. For a good number of these issues they are either forgotten, fixed or not pursued. The rest of them we reopen them manually since they are indeed issues we need to address.

**Comment by niden** [comment]
Some input please. Using the original example

```php
$tag1 = \Tag::findFirst('id=1');
$tag2 = \Tag::findFirst('id=2');
$tag3 = \Tag::findFirst('id=3');

// create a post
$post = new \Post();
$post->tags = [$tag1, $tag2];
$post->save();
// Post is related to tag1 and tag2

// edit a post
$post = \Post::findFirst('id=1'); // the same post that was created earlier
$post->tags = [$tag1, $tag3];
$post->save();
// Post is related to tag1 and tag3
```

This is the intended behavior. If we were to take your approach and the edit action would append records, how would you be able to delete a relationship using edit? For instance based on your suggestion, the proposed functionality will give us:

```php
// edit a post
$post = \Post::findFirst('id=1'); // the same post that was created earlier
$post->tags = [$tag1, $tag3];
$post->save();
// Post is related to tag1, tag2 and tag3
```
So the question is, how would you delete say `tag1` from the relationship and "replace" it by say `tag4` ? If we do not use the direct assignment of what you need to save, then we would have to make an extra call to first delete what we do not want and then assign what we do.

Thoughts on this?

@prodigga @ghost @lfbittencourt @Pajamaman @fl0pp @Jurigag @makerlabs @angelvega93 @StudioMaX @firstactivemedia

**Comment by niden** [comment]
Closing in favor of #13855. Will revisit if the community votes for it, or in later versions.

**Comment by StudioMaX** [comment]
Bump

**Comment by jturbide** [comment]
This is what I am planning to do, the keepMissingRelationship toggle is bothering me. I already have a similar approach working in Phalcon 3.4. I am currently upgrading it to phalcon 4.

**Let's say**
- handling lots of relationships
- recursively saving nested entities indefinately
- belongsTo, hasOne, hasMany, hasManyToMany should behave the same
- matching using multi-field support for relationship binding + forced find condition
- handling model->assign (phalcon 4) or model->save (phalcon 3)
- allowing data manipulation for hasManyToMany nodes as well
- allowing whitelist and getColumns to be passed recursively for aliases as well
- Soft-delete friendlyness on relationships nodes
- Avoiding deleting or re-creating nodes, they could contain some persistant data as well (i.e position)
- Allowing many-to-many node data manipulation
- Allowing to pass a flag if you want to delete missing relationship or keep them

**Saving?**
- beging transaction
- save current model (we have to do it here because the primary keys could possibly be changed)
- get the relationships properties and values using the aliases
	- find their primary keys (can we have combined keys in phalcon?)
		- if primary key is not empty, fetch that entry
			- if soft-delete behaviour, force soft-delete field to the softDelete->notDeleted value
		- if entry can't be found, create new entry
		- assign updated data and nested relationships to that entry
	- if "keepMissingRelationships" flag is false
		- delete missing entries		
	- save and automagically save nested relationships recursively)
	- forward children messages to parent model
		- use alias to prefix fields
- if error, rollback, commit otherwise
	
**Assigning**
```php
        $tagList = [];
	$tagList []= true; // keep missing relationship flag could be passed here if we wanted to have different bahaviour at different nested levels
	$tagList []= Tag::findById(1); // append or replace
	$tagList []= ['name' => 'Tag 2']; // create and append
	$tagList []= ['name' => 'Tag 3', 'PostNode' => ['position' => 3, 'Post' => ['name' => 'Post Name'], 'Tag' => ['name' => 'Tag 3!']]]; // create and append and then update node position
	$tagList []= Tag::findById(4)->assign(['PostNode' => ['position' => 4]]);
	
	$post->TagList = $tagList; // first way
	
	$whiteList = ['TagList' => ['name' => true, 'PostNode' => ['position' => true]]];
	$post->assign(['TagList' => $tagList], $whiteList); // second way, using nested whitelist
	
	$keepMissingRelationships = true;
	$model->save($keepMissingRelationships); // actually do shits using begin rollback & commits, also forward sub getMessages() to their parent model up to the main one
```

**Comment by jturbide** [comment]
https://github.com/zemit-cms/core/blob/master/src/Mvc/Model/Relationship.php
